subprojects/core/src/main/java/org/gradle/internal/classpath/CallInterceptionClosureInstrumentingClassVisitor.java

```diff
@@ -112,25 +113,50 @@ public MethodVisitorScopeImpl(MethodVisitor methodVisitor) {
                     public void visitCode() {
                         /*
                          * enterInstrumentedClosure(this);
-                         * doCall$original(<args>);
-                         * leaveInstrumentedClosure(this);
+                         * try {
+                         *     return doCall$original(<args>);
+                         * } finally { // but with inlining the handler
+                         *     leaveInstrumentedClosure(this);
+                         * }
```

Comment:
❔  My ASM is rusty, but as far as I can tell we are generating something like this:

```groovy
def r;
try {
    r = doCall$original(<args>);
} catch (Throwable t) {
    leaveInstrumentedClosure(this);
    throw t;
}

leaveInstrumentedClosure(this);
return r;
```

If that's correct, wouldn't it better keep this more explicit version of the generated code in the docs?

Is there a way to instrument this without generating two calls to `leaveInstrumentedClosure` ? Or is it something the Java Compiler itself does for `finally` clauses?

Reply:
>Or is it something the Java Compiler itself does for finally clauses?

AFAIR, there's no such thing as `finally` for the JVM, the compiler indeed replicate the `finally` block in all normal returns and catch blocks.

Reply:
Yes, the Java compilers normally inline the `finally` blocks at the corresponding exit points. 
I thought that having the `// but with inlining the handler` comment is enough to indicate that the finally block is inlined while keeping the intent clear enough.

Reply:
I have added some more clarifying comments to this method.